### PR TITLE
New PR for the new web stack

### PR DESCRIPTION
**Pull request by baulig**
[System]: Async-ify and mostly rewrite HttpWebRequest and friends.

`HttpWebRequest` and its helper classes (`WebConnection`, `WebConnectionStream`) have received a major overhaul.
- The old Begin/End async based implementation has been rewritten to use async tasks.  All public API entry points using the APM are now wrappers around their TPL counterparts.
- `WebConnectionStream` has been split into `WebRequestStream` and `WebResponseStream`.  All the reading and writing functionality that was previously spread across `WebConnection`, `WebConnectionStream` and `HttpWebRequest` now lives in these two new classes.
- Each `HttpWebRequest` instance can transparently do more than one actual request due to redirection and / or authentication.  These underlying requests are now called `WebOperation` and each instance of this new class handles exactly one network request.
- The new `WebOperation` class also tracks state regarding sending / receiving / closing streams.
- Most of the old functionality from `WebConnection` has been moved to other classes.  `WebConnection` is now responsible for connecting to the remote server and keeping track of that connection.
- All the connection reuse / idle time / queuing logic has been moved into a new class called `ServicePointScheduler`, replacing `WebConnectionGroup` and the relevant code in `ServicePoint`.
- `ServicePointScheduler` is fully async and keeps track of all currently open connections, all currently running or queued requests and idle connections.  It will automatically clean itself up and exit when it's no longer used.
- Added error checking, and timeouts.
- Fixed content resend logic for `HttpClient`; we now support redirecting POST requests and NTLM Authentication in `HttpClient` using the `HttpContent.CopyToAsync ()` logic.

There were two main objectives for this patch: fixing https://bugzilla.xamarin.com/show_bug.cgi?id=55872 and various issues that random people reported regarding nuget restore printing random error messages - while also modernizing the code-base in a way where it's easier to read, understand and maintain.

This code is very old, most of it was written in 2004 and 2012, and it was using a lot of chained APM async methods which were very hard to understand, debug and fix.  The underlying issue in the bug was that some "EndSomething" was called after the request had been canceled and the connection reused, but some of the code didn't check whether the underlying state had changed.

### CODE REVIEW

File: mcs/class/System/Makefile [hunk 1]
```diff
@@ -30,6 +30,11 @@ TEST_RESOURCES = \
 XTEST_LIB_REFS = System System.Core Facades/System.Threading.Tasks Facades/System.Runtime.InteropServices.RuntimeInformation
 LIB_MCS_FLAGS = -d:CONFIGURATION_2_0 $(REFERENCE_SOURCES_FLAGS) -unsafe $(RESOURCE_FILES:%=-resource:%) -nowarn:436
 
+###
+### FIXME: this is for testing only!
+###
+LIB_MCS_FLAGS += -d:MARTIN_WEB_DEBUG
```
Comment (baulig) [inline mcs/class/System/Makefile]: DEBUGGING

File: mcs/class/System/System.Net.Sockets/Socket.cs [hunk 2]
```diff
@@ -91,7 +91,10 @@ public partial class Socket : IDisposable
 		int m_IntCleanedUp;
 		internal bool connect_in_progress;
 
-#region Constructors
+		static int nextId;
```
Comment (baulig) [inline mcs/class/System/System.Net.Sockets/Socket.cs]: Make conditional to `MONO_WEB_DEBUG` (the field is accessed by debugging code).

**Comment by baulig** [comment]
TODO: rename `MARTIN_WEB_DEBUG` into `MONO_WEB_DEBUG`.

File: mcs/class/System/Test/System.Net/HttpWebRequestTest.cs [hunk 3]
```diff
@@ -303,7 +303,8 @@ public void BeginGetRequestStream_Body_NotAllowed ()
 				request.Method = "GET";
 
 				try {
-					request.BeginGetRequestStream (null, null);
+					var result = request.BeginGetRequestStream (null, null);
+					request.EndGetRequestStream (result);
```
Comment (baulig) [inline mcs/class/System/Test/System.Net/HttpWebRequestTest.cs]: The test was broken, there is on guarantee that `Begin*()` will actually do something synchronously, you should always call the `End*()` method as well to get the exception.
In the new code, `Begin*()` and `End*()` are just wrappers around `*Async()`.

File: mcs/class/System/Test/System.Net/HttpWebRequestTest.cs [hunk 4]
```diff
@@ -3076,7 +3079,8 @@ public void CanWrite ()
 				try {
 					Assert.IsTrue (rs.CanWrite, "#1");
 					rs.Close ();
-					Assert.IsFalse (rs.CanWrite, "#2");
+					// CanRead and CanWrite do not change status after closing.
+					Assert.IsTrue (rs.CanWrite, "#2");
```
Comment (baulig) [inline mcs/class/System/Test/System.Net/HttpWebRequestTest.cs]: Tested behavior was inconsistent with .NET.

File: mcs/class/referencesource/System/net/System/Net/webclient.cs [hunk 5]
```diff
@@ -1805,20 +1805,9 @@ private string GetStringUsingEncoding(WebRequest request, byte[] data)
             OnOpenReadCompleted((OpenReadCompletedEventArgs)arg);
         }
         private void OpenReadAsyncCallback(IAsyncResult result) {
-#if MONO
```
Comment (baulig) [inline mcs/class/referencesource/System/net/System/Net/webclient.cs]: This section has been cleaned up to no longer depend on implementation internals.

File: mcs/class/System/System.Net/HttpWebRequest.cs [hunk 6]
```diff
@@ -853,71 +882,61 @@ public override IAsyncResult BeginGetRequestStream (AsyncCallback callback, obje
...
-			lock (locker)
-			{
+			WebOperation operation;
+			lock (locker) {
 				if (getResponseCalled)
 					throw new InvalidOperationException ("The operation cannot be performed once the request has been submitted.");
 
-				if (asyncWrite != null) {
+				operation = currentOperation;
+				if (operation != null && operation.WriteStream != null)
```
Comment (marek-safar) [inline mcs/class/System/System.Net/HttpWebRequest.cs]: Could be ?.

File: mcs/class/System/System.Net/HttpWebRequest.cs [hunk 7]
```diff
@@ -835,11 +832,43 @@ void AddRange (string rangeSpecifier, long from, long to)
...
+				requestSent = true;
+				if (!redirecting)
+					redirects = 0;
+				servicePoint = GetServicePoint ();
+				servicePoint.SendRequest (operation, connectionGroup);
+				return operation;
+			}
+		}
+
+		Task<Stream> MyGetRequestStreamAsync ()
```
Comment (marek-safar) [inline mcs/class/System/System.Net/HttpWebRequest.cs]: It'd be better to inline this, it's used from a single place anyway

File: mcs/class/System/System.Net/HttpWebRequest.cs [hunk 8]
```diff
@@ -853,71 +882,61 @@ public override IAsyncResult BeginGetRequestStream (AsyncCallback callback, obje
...
+					gotRequestStream = true;
+					operation = SendRequest (false, null, cancellationToken);
 				}
-				return result;
+			}
+
+			try {
+				return await operation.GetRequestStream ().ConfigureAwait (false);
+			} finally {
+				nestedRequestStream = 0;
```
Comment (marek-safar) [inline mcs/class/System/System.Net/HttpWebRequest.cs]: Could you add a comment why
Can this be racy?

File: mcs/class/System/System.Net/HttpWebRequest.cs [hunk 9]
```diff
@@ -926,30 +945,34 @@ public Stream GetRequestStream (out TransportContext context)
...
-			if (contentLength < 0 && writeStream.CanWrite == true && writeStream.WriteBufferLength < 0)
-				return false;
-
-			if (contentLength < 0 && writeStream.WriteBufferLength >= 0)
-				InternalContentLength = writeStream.WriteBufferLength;
-
-			// This will write the POST/PUT if the write stream already has the expected
-			// amount of bytes in it (ContentLength) (bug #77753) or if the write stream
-			// contains data and it has been closed already (xamarin bug #1512).
+			using (var cts = new CancellationTokenSource ()) {
```
Comment (marek-safar) [inline mcs/class/System/System.Net/HttpWebRequest.cs]: Why do you need all this when you could do `task.Wait (timeout)` ?

File: mcs/class/System/System.Net/WebOperation.cs [hunk 10]
```diff
@@ -0,0 +1,342 @@
...
+				return writeStream;
+			}
+		}
+
+		public Task<WebResponseStream> GetResponseStream ()
+		{
+			return responseTask.Task;
+		}
+
+		internal async Task<(bool, WebOperation)> WaitForCompletion (bool ignoreErrors)
```
Comment (marek-safar) [inline mcs/class/System/System.Net/WebOperation.cs]: Where is this used?

File: mcs/class/System/System.Net/WebOperation.cs [hunk 11]
```diff
@@ -0,0 +1,342 @@
...
+
+		public BufferOffsetSize WriteBuffer {
+			get;
+		}
+
+		public bool IsNtlmChallenge {
+			get;
+		}
+
+		static int nextID;
```
Comment (marek-safar) [inline mcs/class/System/System.Net/WebOperation.cs]: Why is this needed?

File: mcs/class/System/System.Net/WebOperation.cs [hunk 12]
```diff
@@ -0,0 +1,342 @@
...
+		{
+			var (exception, disposed) = SetDisposed (ref field);
+			if (disposed)
+				cts?.Cancel ();
+			exception.Throw ();
+		}
+
+		internal void RegisterRequest (ServicePoint servicePoint, WebConnection connection)
+		{
+			lock (this) {
```
Comment (marek-safar) [inline mcs/class/System/System.Net/WebOperation.cs]: Do you really need to lock arguments checks?

File: mcs/class/System/System.Net/WebRequestStream.cs [hunk 13]
```diff
@@ -0,0 +1,436 @@
...
+		bool allowBuffering;
+		bool sendChunked;
+		TaskCompletionSource<int> pendingWrite;
+		long totalWritten;
+		byte[] headers;
+		bool headersSent;
+		int completeRequestWritten;
+		int chunkTrailerWritten;
+
+		internal string ME {
```
Comment (marek-safar) [inline mcs/class/System/System.Net/WebRequestStream.cs]: Should be under debug mode

File: mcs/class/System/System.Net/WebRequestStream.cs [hunk 14]
```diff
@@ -0,0 +1,436 @@
...
+				else
+					Request.InternalContentLength = Operation.WriteBuffer.Size;
+			}
+
+			await SetHeadersAsync (false, cancellationToken).ConfigureAwait (false);
+
+			Operation.ThrowIfClosedOrDisposed (cancellationToken);
+
+			if (Operation.WriteBuffer != null && !Operation.IsNtlmChallenge) {
+				await WriteRequestAsync (cancellationToken);
```
Comment (marek-safar) [inline mcs/class/System/System.Net/WebRequestStream.cs]: Add .ConfigureAwait (false)

File: mcs/class/System/System.Net/WebRequestStream.cs [hunk 15]
```diff
@@ -0,0 +1,436 @@
...
+
+			await SetHeadersAsync (true, cancellationToken).ConfigureAwait (false);
+
+			WebConnection.Debug ($"{ME} WRITE REQUEST #1: {buffer != null}");
+
+			Operation.ThrowIfClosedOrDisposed (cancellationToken);
+			if (buffer != null && buffer.Size > 0)
+				await InnerStream.WriteAsync (buffer.Buffer, 0, buffer.Size, cancellationToken);
+
+			await FinishWriting (cancellationToken);
```
Comment (marek-safar) [inline mcs/class/System/System.Net/WebRequestStream.cs]: Add .ConfigureAwait (false)

File: mcs/class/System/System.Net/WebRequestStream.cs [hunk 16]
```diff
@@ -0,0 +1,436 @@
...
+				throw throwMe;
+			}
+
+			await SetHeadersAsync (true, cancellationToken).ConfigureAwait (false);
+
+			WebConnection.Debug ($"{ME} WRITE REQUEST #1: {buffer != null}");
+
+			Operation.ThrowIfClosedOrDisposed (cancellationToken);
+			if (buffer != null && buffer.Size > 0)
+				await InnerStream.WriteAsync (buffer.Buffer, 0, buffer.Size, cancellationToken);
```
Comment (marek-safar) [inline mcs/class/System/System.Net/WebRequestStream.cs]: Add .ConfigureAwait (false)

File: mcs/class/System/System.Net/WebRequestStream.cs [hunk 17]
```diff
@@ -0,0 +1,436 @@
...
+						break;
+					var ret = await Task.WhenAny (timeoutTask, oldTcs.Task).ConfigureAwait (false);
+					if (ret == timeoutTask)
+						throw new WebException ("The operation has timed out.", WebExceptionStatus.Timeout);
+				}
+
+				try {
+					await WriteChunkTrailer_inner (cts.Token).ConfigureAwait (false);
+				} catch {
+					;
```
Comment (marek-safar) [inline mcs/class/System/System.Net/WebRequestStream.cs]: Why are we hiding exceptions here?

File: mcs/class/System/System.Net/WebRequestStream.cs [hunk 18]
```diff
@@ -0,0 +1,436 @@
...
+		{
+			using (var cts = new CancellationTokenSource ()) {
+				cts.CancelAfter (WriteTimeout);
+				var timeoutTask = Task.Delay (WriteTimeout);
+				while (true) {
+					var myWriteTcs = new TaskCompletionSource<int> ();
+					var oldTcs = Interlocked.CompareExchange (ref pendingWrite, myWriteTcs, null);
+					if (oldTcs == null)
+						break;
+					var ret = await Task.WhenAny (timeoutTask, oldTcs.Task).ConfigureAwait (false);
```
Comment (marek-safar) [inline mcs/class/System/System.Net/WebRequestStream.cs]: Again why not `oldTcs.Task.Wait (WriteTimeout)` ?

File: mcs/class/System/System.Net/HttpWebRequest.cs (hunk 9 above)
Comment (baulig) [reply mcs/class/System/System.Net/HttpWebRequest.cs]: We need two things:
a) A `CancellationToken` that's automatically being cancelled after the timeout (with the `abort` delegate being called).
b) Return a Task that is guaranteed to complete within the timeout.

So using `task.Wait (timeout)` here doesn't help us here.

However, we don't have to actually pass the `timeoutTask` to the inner method, I'm going to fix that.

File: mcs/class/System/System.Net/WebRequestStream.cs (hunk 16 above)
Comment (baulig) [reply mcs/class/System/System.Net/WebRequestStream.cs]: There is already a `.ConfigureAwait (false)` in this method that will be called unconditionally before we get here, do we really need it on each `await`?

File: mcs/class/System/System.Net/WebRequestStream.cs (hunk 18 above)
Comment (baulig) [reply mcs/class/System/System.Net/WebRequestStream.cs]: Apart from this being an async method (and I once learned that you should not sync block in it) - this would also change semantics here.

It is possible that whatever is awaiting on `WriteAsync ()` wakes up before we do and starts another write, so we'd actually loop here - but we want to wait a maximum of `WriteTimeout`, so therefor the timer is initialized outside the loop.

File: mcs/class/System/System.Net/WebRequestStream.cs (hunk 17 above)
Comment (baulig) [reply mcs/class/System/System.Net/WebRequestStream.cs]: Intentional, going to stay.  I'll mark it in code.  This is very broken anyways.

Comment (baulig) [reply mcs/class/System/System.Net/WebRequestStream.cs]: See comment (and FIXME) in line 408.

File: mcs/class/System/System.Net/HttpWebRequest.cs (hunk 8 above)
Comment (baulig) [reply mcs/class/System/System.Net/HttpWebRequest.cs]: I believe this variable does not need to be reset, but need to check this.
**MARTIN TODO**

File: mcs/class/System/System.Net/HttpWebRequest.cs (hunk 7 above)
Comment (baulig) [reply mcs/class/System/System.Net/HttpWebRequest.cs]: Done.

File: mcs/class/System/System.Net/HttpWebRequest.cs (hunk 6 above)
Comment (baulig) [reply mcs/class/System/System.Net/HttpWebRequest.cs]: Done

File: mcs/class/System/System.Net/HttpWebRequest.cs (hunk 9 above)
Comment (baulig) [reply mcs/class/System/System.Net/HttpWebRequest.cs]: Cleaned this up, is it better now?

File: mcs/class/System/System.Net/WebOperation.cs (hunk 12 above)
Comment (baulig) [reply mcs/class/System/System.Net/WebOperation.cs]: Moved argument checks out.

File: mcs/class/System/System.Net/WebOperation.cs (hunk 10 above)
Comment (baulig) [reply mcs/class/System/System.Net/WebOperation.cs]: ServicePointScheduler.cs, line 163

File: mcs/class/System/System.Net/HttpWebRequest.cs (hunk 9 above)
Comment (marek-safar) [reply mcs/class/System/System.Net/HttpWebRequest.cs]: there is following override for exactly that
```
public bool Wait(
	int millisecondsTimeout,
	CancellationToken cancellationToken
)
```

Comment (marek-safar) [reply mcs/class/System/System.Net/HttpWebRequest.cs]: I still don't understand why you need to set up 2 identical cancellations. You setup timeoutthe cancellation timeout which will throw (and you are missing catch for this) and extra wait on timeout.

I think simple 
```
try {
await func (timeoutTask, cts.Token).ConfigureAwait (false);
} catch (OperationCanceledException) {
}
```
should work or if you think the cancellation is not imminent, don't set it with timeout and Cancell it manually when you get timeoutTask result.

**Comment by refic** [comment]
Hi folks,

I've been testing this with a web crawler, however it always crashes after running for a few minutes with:

[12/1/2017 9:20:14 AM] Unhandled exception caught: System.NullReferenceException: Object reference not set to an instance of an object
  at Majestic12.WebCrawler.AsyncBeginGetResponseStream (System.IAsyncResult oAR) [0x002f3] in <a0575f6a05884103be8913cdd4b349ce>:0
  at System.Threading.Tasks.TaskToApm+<>c__DisplayClass3_0.<InvokeCallbackWhenTaskCompletes>b__0 () [0x00000] in <3b72029063bc4ca795c071cc8fe419df>:0
  at System.Threading.Tasks.AwaitTaskContinuation.InvokeAction (System.Object state) [0x00000] in <3b72029063bc4ca795c071cc8fe419df>:0
  at System.Threading.ExecutionContext.RunInternal (System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, System.Object state, System.Boolean preserveSyncCtx) [0x00071] in <3b72029063bc4ca795c071cc8fe419df>:0
  at System.Threading.ExecutionContext.Run (System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, System.Object state, System.Boolean preserveSyncCtx) [0x00000] in <3b72029063bc4ca795c071cc8fe419df>:0
  at System.Threading.Tasks.AwaitTaskContinuation.RunCallback (System.Threading.ContextCallback callback, System.Object state, System.Threading.Tasks.Task& currentTask) [0x0001a] in <3b72029063bc4ca795c071cc8fe419df>:0
--- End of stack trace from previous location where exception was thrown ---
  at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw () [0x0000c] in <3b72029063bc4ca795c071cc8fe419df>:0
  at System.Threading.Tasks.AwaitTaskContinuation+<>c.<ThrowAsyncIfNecessary>b__17_0 (System.Object s) [0x00000] in <3b72029063bc4ca795c071cc8fe419df>:0
  at System.Threading.QueueUserWorkItemCallback.WaitCallback_Context (System.Object state) [0x00007] in <3b72029063bc4ca795c071cc8fe419df>:0
  at System.Threading.ExecutionContext.RunInternal (System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, System.Object state, System.Boolean preserveSyncCtx) [0x00071] in <3b72029063bc4ca795c071cc8fe419df>:0
  at System.Threading.ExecutionContext.Run (System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, System.Object state, System.Boolean preserveSyncCtx) [0x00000] in <3b72029063bc4ca795c071cc8fe419df>:0
  at System.Threading.QueueUserWorkItemCallback.System.Threading.IThreadPoolWorkItem.ExecuteWorkItem () [0x00021] in <3b72029063bc4ca795c071cc8fe419df>:0
  at System.Threading.ThreadPoolWorkQueue.Dispatch () [0x00074] in <3b72029063bc4ca795c071cc8fe419df>:0
  at System.Threading._ThreadPoolWaitCallback.PerformWaitCallback () [0x00000] in <3b72029063bc4ca795c071cc8fe419df>:0

Do let me know if I can be of help with this. Source code for the crawler is not available though.

**Comment by baulig** [comment]
Cleaned up stack trace from the bug report:
 
```
  at WebException..ctor (string message, string data, exception innerException,
      WebExceptionStatus status, WebResponse response) [0x0005a] in
      referencesource/System/net/System/Net/WebException.cs:141
  at WebException..ctor (string message, exception innerException, WebExceptionStatus status,
      WebResponse response) [0x00000] in
      referencesource/System/net/System/Net/WebException.cs:129
  at HttpWebRequest.GetWebException (System.Excep111tion e) [0x0004d] in
      System/System.Net/HttpWebRequest.cs:1146
  at HttpWebRequest+<MyGetResponseAsync>d__239.MoveNext () [0x0038f] in
      System/System.Net/HttpWebRequest.cs:1012
  at WebOperation.SetError (exception error) [0x0001a] in
      System/System.Net/WebOperation.cs:146
  at WebOperation+<Run>d__54.MoveNext () [0x00213] in
      System/System.Net/WebOperation.cs:279
  at WebResponseStream+<InitReadAsync>d__61.MoveNext () [0x002e7] in
      System/System.Net/WebResponseStream.cs:617
  at Sockets.SocketAsyncResult+<>c.<Complete>b__27_0 (System.Object state) [0x00000] in
      System/Sockets/SocketAsyncResult.cs:157
  at QueueUserWorkItemCallback.IThreadPoolWorkItem.ExecuteWorkItem () [0x00015] in
      referencesource/mscorlib/system/threading/threadpool.cs:1279

  at WebException.get_Response () [0x00036] in
      referencesource/System/net/System/Net/WebException.cs:216
  at Majestic12.WebCrawler.AsyncBeginGetResponseStream (System.IAsyncResult oAR) [0x002ea] in
      <a0575f6a05884103be8913cdd4b349ce>:0
```

**Comment by baulig** [comment]
@monojenkins build pkg

**Comment by baulig** [comment]
@monojenkins build pkg

**Comment by marek-safar** [comment]
build

**Comment by baulig** [comment]
build
